### PR TITLE
Fix weaver generate process when routing key of Router component is a struct

### PR DESCRIPTION
**Pull request by zailic**
More details in https://github.com/ServiceWeaver/weaver/issues/414 issue.

### CODE REVIEW

**Review by mwhittaker — COMMENTED** [review]
Amazing job! Would you mind updating the documentation of `isValidRouterType` and also the documentation of valid router key types in `component.go` and `website/docs.md`? You can also leave that to me if you'd like :)

**Comment by zailic** [comment]
Thanks for your quick feedback! I just updated documentation as per your request. Feel free to refine it as my English is not as good as yours! :D

**Review by mwhittaker — APPROVED** [review]
Awesome, thanks so much for your contribution!

**Comment by mwhittaker** [comment]
> Thanks for your pull request! It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> View this [failed invocation](https://github.com/ServiceWeaver/weaver/pull/415/checks?check_run_id=14538281254) of the CLA check for more information.
> 
> For the most up to date status, view the checks section at the bottom of the pull request.

@zailic I think you'll have to complete these instructions before I can merge in the PR.

File: internal/tool/generate/types.go
Comment (mwhittaker) [inline internal/tool/generate/types.go:1]: Oh, I actually think your original approach is correct. A router key type does not need to embed AutoMarshal; it is never actually marshaled. But, ignoring an embedded `weaver.AutoMarshal` makes sense to me.

Comment (zailic) [reply internal/tool/generate/types.go:1]: Hmm, then things are more complicated than I thought 😆. Ultimately the router key should be passed as an input argument of one of the method of a weaver component which is associated with the router. Currently the `weaver generate process` requires that the methods of weaver components should accept only serializable arguments - https://github.com/ServiceWeaver/weaver/blob/main/internal/tool/generate/generator.go#L666-L675 . Therefore to satisfy the requirement above I think it's mandatory that the router key does need to embed AutoMarshal

Comment (mwhittaker) [reply internal/tool/generate/types.go:1]: Ah, I think I see the confusion. Router keys do not have to be method arguments. For example, we could re-write your example in #414 to the following: 


```go
type Cache interface {
    Get(ctx context. Context, namespace, key string) ([]string, error)	
    Put(ctx context.Context, namespace, key string, emojis []string) error
}

// Cache implementation...

type router struct{}

type routingKey struct {
    namespace string
    key string
}

func (router) Get(_ context.Context, namespace, key string) routingKey {
	return routingKey{namespace, key}
}

func (router) Put(_ context.Context, namespace, key string, _ []string) routingKey {
	return routingKey{namespace, key}
}
```

Comment (mwhittaker) [reply internal/tool/generate/types.go:1]: But I think your example made a great point that it's probably not uncommon to use a method argument as a routing key, and allowing `weaver.AutoMarshal` in keys is easy and relatively intuitive, so I think it makes sense to support it.

Comment (zailic) [reply internal/tool/generate/types.go:1]: Ahh, you're right...I should have read the documentation more carefully. Thanks for the clarification 🙏!

**Comment by mwhittaker** [comment]
Hey @zailic, mind squashing and rebasing on the main branch?

**Pull request merged**
